File: mpi/integrate2.f90
program main
   use MPI
   implicit none
   integer :: size, rank, i, ierror
   double precision :: a, b, res, mya, myb, psum
   double precision, parameter :: pi = 4.0d0 * atan(1.0d0)
   integer, allocatable, dimension(:,:) :: statuses
   integer, allocatable, dimension(:) :: requests
   double precision, allocatable, dimension(:) :: tmp
   double precision, external :: integrate

   call MPI_Init(ierror)
   call MPI_Comm_size(MPI_COMM_WORLD, size, ierror)
   call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)

   ! integration limits
   a = 0.0d0 ; b = 2.0d0 * pi ; res = 0.0d0

   if(rank.eq.0) then
      allocate(statuses(MPI_STATUS_SIZE, size-1))
      allocate(requests(size-1))
      allocate(tmp(size-1))
      ! pre-post nonblocking receives
      do i=1,size-1
         call MPI_Irecv(tmp(i), 1, MPI_DOUBLE_PRECISION, &
                        i, 0, MPI_COMM_WORLD,       &
                        requests(i), ierror)
      enddo
   endif
   ! limits for "me"
   mya = a + rank*(b-a)/size
   myb = mya + (b-a)/size

   ! integrate f(x) over my own chunk - actual work
   psum = integrate(mya,myb)

   ! rank 0 collects partial results
   if(rank.eq.0) then
      res = psum
      call MPI_Waitall(size-1, requests, statuses, ierror)
      do i=1,size-1
         res = res + tmp(i)
      enddo
      write (*,*) 'Result: ',res
   ! ranks != 0 send their results to rank 0
   else
      call MPI_Send(psum, 1, &
                    MPI_DOUBLE_PRECISION, 0, 0, &
                    MPI_COMM_WORLD,ierror)
   endif

   call MPI_Finalize(ierror)
end program main

! Return integral of cos(x) from x=a to x=b
double precision function integrate(a, b)
   implicit none
   double precision :: a, b
   integrate = sin(b) - sin(a)
end function integrate

File: mpi/integrate2.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

// Return integral of cos(x) from x=a to x=b
double integrate(double a, double b)
{
   return sin(b) - sin(a);
}

// Compute integral(x=a to x=b) f(x)
int main(int argc, char** argv)
{
   int size, rank, i, ierr;
   double a, b, res, mya, myb, psum, *tmp;
   MPI_Request *request;
   MPI_Status  *status;

   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &size);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   // integration limits
   a = 0.0; b = 2.0; res = 0.0;

   // Start non-blocking recives on rank=0
   if(rank == 0)
   {
      status  = (MPI_Status*)malloc(size*sizeof(MPI_Status));
      request = (MPI_Request*)malloc(size*sizeof(MPI_Request));
      tmp     = (double*)malloc(size*sizeof(double));
      for(i=1; i<size; ++i)
         MPI_Irecv(&tmp[i], 1, MPI_DOUBLE, i, 0, MPI_COMM_WORLD, &request[i]);
   }

   // limits for "me"
   mya = a + rank*(b-a)/size;
   myb = mya + (b-a)/size;
   // integrate f(x) over my own chunk - actual work
   psum = integrate(mya, myb);

   // rank 0 collects partial results
   if(rank == 0)
   {
      res = psum;
      MPI_Waitall(size-1, &request[1], &status[1]);
      for(i=1; i<size; ++i)
         res += tmp[i];
      printf("Result: %20.14e\n",res);
   }
   // ranks != 0 send their results to rank 0
   else
   {
      MPI_Send(&psum,                // send buffer
               1,                    // message length
               MPI_DOUBLE_PRECISION,
               0,                    // rank of destination
               0,                    // tag (unused here)
               MPI_COMM_WORLD);
   }

   if(rank == 0)
   {
      free(request); free(status); free(tmp);
   }
   MPI_Finalize();
}
